### PR TITLE
luna: add table reader

### PR DESCRIPTION
**Pull request by w41ter-l**
This PR ref #221

### CODE REVIEW

**Comment by huachaohuang** [comment]
Thanks, I will take a look later today.

**Review by zojw — DISMISSED** [review]

File: src/engine/luna/src/table/table_iter.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            if self.index_iter.valid() {
+                let mut iter = read_block_iter(&mut self.reader, self.index_iter.value()).await?;
+                iter.seek_to_first();
+                self.block_iter = Some(iter);
+            } else {
+                self.block_iter = None;
+            }
+        }
+
+        Ok(())
```
Comment (zojw) [inline src/engine/luna/src/table/table_iter.rs:96]: BTW: when do we need to return `Err(..)`~? it seems current we need `.valid` after `next` :)

Comment (w41ter-l) [reply src/engine/luna/src/table/table_iter.rs:96]: `next` returns `Err` if it meets IO error. In this function, `read_block_iter` might returns `Err`.

Comment (zojw) [reply src/engine/luna/src/table/table_iter.rs:96]: thanks got it... I miss `?`  orz

File: src/engine/luna/src/table/table_reader.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        let index_block_handler = BlockHandle::decode_from(&mut footer.as_slice());
+        let block_content = read_block(&mut reader, &index_block_handler).await?;
+
+        Ok(TableReader {
+            reader,
+            index_block: BlockReader::new(block_content),
+        })
+    }
+
+    pub fn iter(&mut self) -> TableIter<R> {
```
Comment (huachaohuang) [inline src/engine/luna/src/table/table_reader.rs:46]: I think `TableReader` should be thread-safe and doesn't require a mutable reference to read. I created a related issue [here](https://github.com/engula/engula/issues/261). Maybe you can help to resolve it?

Comment (w41ter-l) [reply src/engine/luna/src/table/table_reader.rs:46]: I'll fix that problem first and then update this PR.

Comment (huachaohuang) [reply src/engine/luna/src/table/table_reader.rs:47]: I think this is not solved?

Comment (w41ter-l) [reply src/engine/luna/src/table/table_reader.rs:47]: Oh, I missed this part.

Comment (w41ter-l) [reply src/engine/luna/src/table/table_reader.rs]: @huachaohuang already updated.

**Review by huachaohuang — APPROVED** [review]
LGTM. Note that I am not concerned about the code quality in the engine part for now.

**Pull request merged**